File: CS/WinWebSolution.Module/Contacts.cs
using System;
using DevExpress.Xpo;
using System.ComponentModel;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;

namespace WinWebSolution.Module {
    [DefaultClassOptions]
    [System.ComponentModel.DisplayName("Contacts")]
    public abstract class ContactBase : Person {
        protected const string DocumentAssignedToContactCriteriaBase = "AssignedTo=?";
        public ContactBase(Session session) : base(session) { }
        [Browsable(false)]
        public virtual CriteriaOperator AssignedDocumentsCriteria {
            get {
                return CriteriaOperator.Parse(DocumentAssignedToContactCriteriaBase, this);
            }
        }
        private bool activeCore;
        public bool Active {
            get { return activeCore; }
            set { SetPropertyValue("Active", ref activeCore, value); }
        }
    }
    public class ContactType1 : ContactBase {
        protected const string DocumentAssignedToContactCriteria1 = "ObjectType.TypeName='WinWebSolution.Module.DocumentType1'";
        public ContactType1(Session session) : base(session) { }
        private string contactType1PropertyCore;
        public string ContactType1Property {
            get { return contactType1PropertyCore; }
            set { SetPropertyValue("ContactType1Property", ref contactType1PropertyCore, value); }
        }
        public override CriteriaOperator AssignedDocumentsCriteria {
            get {
                return base.AssignedDocumentsCriteria & CriteriaOperator.Parse(DocumentAssignedToContactCriteria1);
            }
        }
    }
    public class ContactType2 : ContactBase {
        protected const string DocumentAssignedToContactCriteria2 = "ObjectType.TypeName='WinWebSolution.Module.DocumentType2'";
        public ContactType2(Session session) : base(session) { }
        private string contactType2PropertyCore;
        public string ContactType2Property {
            get { return contactType2PropertyCore; }
            set { SetPropertyValue("ContactType2Property", ref contactType2PropertyCore, value); }
        }
        public override CriteriaOperator AssignedDocumentsCriteria {
            get {
                return base.AssignedDocumentsCriteria & CriteriaOperator.Parse(DocumentAssignedToContactCriteria2);
            }
        }
    }
}

File: VB/WinWebSolution.Module/Contacts.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo
Imports System.ComponentModel
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl

Namespace WinWebSolution.Module
	<DefaultClassOptions, System.ComponentModel.DisplayName("Contacts")> _
	Public MustInherit Class ContactBase
		Inherits Person
		Protected Const DocumentAssignedToContactCriteriaBase As String = "AssignedTo=?"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		<Browsable(False)> _
		Public Overridable ReadOnly Property AssignedDocumentsCriteria() As CriteriaOperator
			Get
				Return CriteriaOperator.Parse(DocumentAssignedToContactCriteriaBase, Me)
			End Get
		End Property
		Private activeCore As Boolean
		Public Property Active() As Boolean
			Get
				Return activeCore
			End Get
			Set(ByVal value As Boolean)
				SetPropertyValue("Active", activeCore, value)
			End Set
		End Property
	End Class
	Public Class ContactType1
		Inherits ContactBase
		Protected Const DocumentAssignedToContactCriteria1 As String = "ObjectType.TypeName='WinWebSolution.Module.DocumentType1'"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private contactType1PropertyCore As String
		Public Property ContactType1Property() As String
			Get
				Return contactType1PropertyCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("ContactType1Property", contactType1PropertyCore, value)
			End Set
		End Property
		Public Overrides ReadOnly Property AssignedDocumentsCriteria() As CriteriaOperator
			Get
				Return MyBase.AssignedDocumentsCriteria And CriteriaOperator.Parse(DocumentAssignedToContactCriteria1)
			End Get
		End Property
	End Class
	Public Class ContactType2
		Inherits ContactBase
		Protected Const DocumentAssignedToContactCriteria2 As String = "ObjectType.TypeName='WinWebSolution.Module.DocumentType2'"
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private contactType2PropertyCore As String
		Public Property ContactType2Property() As String
			Get
				Return contactType2PropertyCore
			End Get
			Set(ByVal value As String)
				SetPropertyValue("ContactType2Property", contactType2PropertyCore, value)
			End Set
		End Property
		Public Overrides ReadOnly Property AssignedDocumentsCriteria() As CriteriaOperator
			Get
				Return MyBase.AssignedDocumentsCriteria And CriteriaOperator.Parse(DocumentAssignedToContactCriteria2)
			End Get
		End Property
	End Class
End Namespace